PhysicsTools/JetMCAlgos/plugins/ttHFGenFilter.cc

```diff
@@ -187,67 +187,60 @@ bool ttHFGenFilter::HasAdditionalBHadron(const std::vector<int>& genBHadIndex, c
 
 //recursive function, that loops over all chain particles  until it finds the protons
 bool ttHFGenFilter::analyzeMothersRecursive(const reco::Candidate* particle,std::vector<const reco::Candidate*>& AllTopMothers){
-  //std::cout << "Particle: " << particle->pdgId() << " , Status: " << particle->status() << " , numberOfMothers: " << particle->numberOfMothers() << std::endl;
   if(particle->status()>20&&particle->status()<30){ //particle comes from hardest process in event
     return true;
   }
   for(uint k=0; k<AllTopMothers.size();k++){
-    if(particle==AllTopMothers[k]){ //partcile comes from  ISR
-
+    if(particle==AllTopMothers[k]){ //particle comes from  ISR
       return true;
     }
   }
-    bool IsFromHardProcess=false;
-  /*if(analyzeMothersRecursive(particle)){
-    return true;
-  }*/
-    for(uint i=0;i<particle->numberOfMothers();i++){
-      //std::cout << "i " <<  i << std::endl;
-      const reco::Candidate* mother = particle->mother(i);
-      IsFromHardProcess=analyzeMothersRecursive(mother,AllTopMothers);
-      if(IsFromHardProcess){
-        return true;
-      }
-
+  bool IsFromHardProcess=false;
+  for(uint i=0;i<particle->numberOfMothers();i++){
+    const reco::Candidate* mother = particle->mother(i);
+    IsFromHardProcess=analyzeMothersRecursive(mother,AllTopMothers);
+    if(IsFromHardProcess){
+      return true;
     }
-
+  }
   return IsFromHardProcess;
 }
 
 std::vector< const reco::Candidate*> ttHFGenFilter::GetTops(const std::vector<reco::GenParticle>& genParticles, std::vector<const reco::Candidate*>& AllTopMothers){
-  const reco::GenParticle* FirstTop = 0;
-  const reco::GenParticle* FirstTopBar = 0;
-  bool FoundTop = false;
-  bool FoundTopBar =false;
+  //  std::vector<const reco::Candidate*> tops;
+  const reco::GenParticle* firstTop = nullptr;
+  const reco::GenParticle* firstTopBar = nullptr;
+  bool foundTop = false;
+  bool foundTopBar = false;
   std::vector<const reco::GenParticle*> Tops;
 
   //loop over all genParticles and find  a Top and AntiTop quark
   //then find all mothers of the Tops
   //this is then used to  if the b-hadron is an inital state radiation particle
 
    for(reco::GenParticleCollection::const_iterator i_particle = genParticles.begin(); i_particle != genParticles.end(); ++i_particle){
-       const reco::GenParticle* thisParticle = &*i_particle;
-     if(thisParticle->pdgId()==6){
-       FirstTop = thisParticle;
-       FoundTop = true;
+     const reco::GenParticle* thisParticle = &*i_particle;
+     firstTop = thisParticle;
+     if(!foundTop && firstTop->pdgId()==6){
```

Comment:
I'm also fixing a bug here: if a second top is found before the first anti-top, then the second top was selected instead of the first one. (The same happened for the anti-top)